GenericList/list.h

```diff
@@ -0,0 +1,260 @@
+#pragma once
+
+#include <iostream>
+#include <stdexcept>
+#include <string>
+#include <memory>
+#include <cassert>
+#include <functional>
+#include <algorithm>
+#include <type_traits>
+
+// Never use header-wide using directives ("using namespace") in the header!!
+// Explanation: https://stackoverflow.com/questions/5849457/using-namespace-in-c-headers
+// Basically, when this header is included, the "using namespace" will be included too, possibly causing un-intended clashes in variable/function/etc.. namess
+// Note: as long as code doesn't #include a .cpp file, it's fine to use using directives in that source file
+
+//Generic type, allow for stateful Allocator if user desires it
+template <typename T, typename Allocator = std::allocator<T>>
+class List {
+    static_assert(!std::is_void_v<T>, "void type is not allowed");
+    static_assert(!std::is_reference_v<T>, "reference type is not allowed");
+    //C++ Standard �8.3.2/4: There shall be no references to references, no arrays of references, and no pointers to references.
+
+
+    //note: elements are value copies of original objects (copy-by-value, not copy-by-reference)
+public:
+    List() : dataAllocator(), data(nullptr), capacity(0), count(0) {
+
+    }
+
+    List(Allocator const& alloc) : dataAllocator(alloc), data(nullptr), capacity(0), count(0) {
+    
+    }
+
+    friend void swap(List& first, List& second) noexcept {
+        //Purely swap data: since we want the states to remain the same as beforehand, just switched, explicitely use std::swap in case swap(Allocator&, Allocator&) somehow moves memory around
+        //Note: since data is a pointer value, it cannot have a "custom swap friend function" - aka, let's just use std::swap everywhere
+        std::swap(first.dataAllocator, second.dataAllocator);
+        std::swap(first.capacity, second.capacity);
+        std::swap(first.count, second.count);
+        std::swap(first.data, second.data);
+    }
+
+    //Rule of 3
+    ~List() { //Destructor
+        Clear();
+        dataAllocator.deallocate(data, capacity);
+    }
+
+    //Copy Constructor - dataAllocator is not copied over as the data allocated is not the same (different memory address, deep copy)
+    List(const List& other) : dataAllocator(), data(CreateDeepCopy(other.data, other.capacity, other.count, dataAllocator)),
```

Comment:
Copy the allocator as well. 

There is a "proper" way of doing it but it's a bit too complicated. The standard library usually uses a set of proxy functions found in `std::allocator_traits` to standardize some of the allocator API calls. For example, in a container copy constructor, they'd call 
```c++
using AllocTraits = std::allocator_traits<Allocator>;
List(const List& other) : dataAllocator(AllocTraits::select_on_container_copy_construction(other.dataAllocator)),
```